packages/validation/src/type-checker/module-context.js

```diff
@@ -0,0 +1,138 @@
+const debug = require("debug")("webassemblyjs:typechecker");
+
+/**
+ * Module context for type checking
+ */
+export default class ModuleContext {
+  constructor() {
+    this.funcs = [];
+    this.globals = [];
+    this.mems = [];
+
+    // Current stack frame
+    this.locals = [];
+    this.labels = [];
+
+    this.debugName = "unknown";
+  }
+
+  /**
+   * Reset the active stack frame
+   */
+  newContext(debugName, expectedResult) {
+    debug("new context %s", debugName);
+
+    this.locals = [];
+    this.labels = [expectedResult];
+    this.debugName = debugName;
+  }
+
+  /**
+   * Functions
+   */
+  addFunction({ params: args, results: result }) {
+    args = args.map(arg => arg.valtype);
+
+    debug("add new function %s -> %s", args.join(" "), result.join(" "));
+
+    this.funcs.push({ args, result });
+  }
+
+  importFunction({ params: args, results: result }) {
+    args = args.map(arg => arg.valtype);
+
+    debug(
+      "add new imported function %s -> %s",
+      args.join(" "),
+      result.join(" ")
+    );
+
+    this.funcs.unshift({ args, result });
+  }
+
+  hasFunction(index) {
+    return typeof this.getFunction(index) !== "undefined";
+  }
+
+  getFunction(index) {
+    if (typeof index !== "number") {
+      throw new Error("getFunction only supported for number index");
```

Comment:
When name are restored from the binary, getFunction is called with an Identifer and the lookup doesn't work. Maybe we should disable name resolution or keep both.